### PR TITLE
Algorithm to calculate pixel grouping calibration state parameters

### PR DESCRIPTION
**Pull request by bolotovskyr**
Creates an algorithm/recipe/service to calculate detector pixel grouping-specific parameters: 2Theta, absolute and relative resolution. The algorithm takes in a calibration state, an instrument definition file and a pixel grouping definition file. 
NOTE:

- this PR updates `environment.yml` and `.github/workflows/actions.yml` to use the `mantid` nightly build
- To test the algorithm, please run `pytest` on the analysis machine

### CODE REVIEW

File: tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        assert len(pixelGroupingParams_ref["dMax"]) == number_of_groupings_calc
+        assert len(pixelGroupingParams_ref["delDOverD"]) == number_of_groupings_calc
+
+        # reverseGroupingIndex takes care of the different order of pixel groupings between "full" anf "lite"
+        # instruments. This has to do with how Mantid treats different kinds of grouping files used by
+        # PixelGroupingParametersCalculationAlgorithm
+        index = 0 if reverseGroupingIndex else number_of_groupings_calc - 1
+        for param in pixelGroupingParams_ref["twoTheta"]:
+            assert abs(float(param) - pixelGroupingParams_calc[index].twoTheta) < 1.0e-5
+            index += 1 if reverseGroupingIndex else -1
```
Comment (rboston628) [inline tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py:226]: `index += (1 if reverseGroupingIndex else -1)` is easier to understand at a glance

File: src/snapred/backend/recipe/algorithm/PixelGroupingParametersCalculationAlgorithm.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        # load the idf into workspace
+        idf = self.getProperty("InstrumentDefinitionFile").value
+
+        # get detector state from the input state
+        calibrationState = Calibration.parse_raw(self.getProperty("InputState").value)
+        detectorState = calibrationState.instrumentState.detectorState
+
+        # add sample logs
+        self.mantidSnapper.AddSampleLog(
```
Comment (rboston628) [inline src/snapred/backend/recipe/algorithm/PixelGroupingParametersCalculationAlgorithm.py:122]: Could this be done in a loop?

File: environment.yml [hunk 3]
```diff
@@ -2,7 +2,7 @@ name: SNAPRed
 channels:
 - conda-forge
 - default
-- mantid
+- mantid/label/nightly
```
Comment (rboston628) [inline environment.yml:5]: This part makes me uncomfortable

Comment (walshmm) [reply environment.yml:5]: Its fine, the plan going forward is to pin versions in qa and main with every release.  Its a bit of a compromise as we cant wait for the mantid release cycle to move forward with these changes.

File: src/snapred/backend/dao/state/PixelGroupingParameters.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from pydantic import BaseModel
+
+from snapred.backend.dao.Limit import Limit
+
+
+class PixelGroupingParameters(BaseModel):
```
[the remark this answers is not in the document]
Comment (bolotovskyr) [reply src/snapred/backend/dao/state/PixelGroupingParameters.py:6]: I called it a state at first, but then realized these are just a couple of auxiliary parameters for calibration, not really a state.

File: src/snapred/backend/dao/PixelGroupingIngredients.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class PixelGroupingIngredients(BaseModel):
+    """Class to hold the ingredients necessary for pixel grouping parameter calculation."""
+
+    calibrationState: Calibration
+    instrumentDefinitionFile: str
+    groupingFile: str
+
+    # if we need specific getter and setter methods, we can use the @property decorator
```
[the remark this answers is not in the document]
Comment (bolotovskyr) [reply src/snapred/backend/dao/PixelGroupingIngredients.py]: Done.

File: tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+import socket
+import unittest.mock as mock
+
+import pytest
+
+with mock.patch.dict(
+    "sys.modules",
+    {
+        # "mantid.api": mock.Mock(),
+        # "mantid.kernel": mock.Mock(),
```
[the remark this answers is not in the document]
Comment (bolotovskyr) [reply tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py]: Done.

File: tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py (hunk 1 above)
Comment (bolotovskyr) [reply tests/unit/backend/recipe/algorithm/test_PixelGroupingParametersCalculationAlgorithm.py:224]: Done.

File: src/snapred/backend/recipe/algorithm/PixelGroupingParametersCalculationAlgorithm.py (hunk 2 above)
Comment (bolotovskyr) [reply src/snapred/backend/recipe/algorithm/PixelGroupingParametersCalculationAlgorithm.py]: Done.

**Review by rboston628 — APPROVED** [review]

**Pull request merged**